lib/validator/task/mod.rs

```diff
@@ -768,101 +772,387 @@ pub(in crate::validator) fn disconnect_block(
+        dbs.block_hashes.put_headers(
+            &mut rwtxn,
+            &headers
+                .iter()
+                .map(|(header, _, height)| (*header, *height))
+                .collect::<Vec<_>>(),
+        )?;
         let () = rwtxn.commit()?;
-        block_hash = latest_missing_header;
+
+        // Send progress update
         let progress = HeaderSyncProgress {
-            current_height: Some(height),
+            current_height: Some(current_height),
         };
-        // Send progress through watch channel
         if let Err(err) = progress_tx.send(progress) {
             tracing::warn!("Failed to send header sync progress: {err:#}");
         }
+
+        // Important: break out here /after/ storing the last header.
+        if current_block_hash == main_tip {
+            tracing::info!("Reached main tip at height {current_height}!");
+            break;
+        }
+    }
+
+    // After finishing the loop, current_height is the height of the active chain.
+    // If the active chain /regressed/ (i.e. `invalidateblock`), we can end up
+    // in a situation where the current enforcer tip is /ahead/ of the active chain,
+    // on an invalid fork.
+    let mut rwtxn = dbs.write_txn()?;
+    let current_enforcer_tip_height = dbs
+        .block_hashes
+        .height()
+        .get(&rwtxn, &current_enforcer_tip)?;
+
+    if current_enforcer_tip_height > current_height {
+        tracing::warn!("Enforcer tip (#{current_enforcer_tip_height}) is ahead of the active chain (#{current_height}), moving back to `{current_block_hash}`");
+        let () = dbs
+            .current_chain_tip
+            .put(&mut rwtxn, &UnitKey, &current_block_hash)?;
+        let () = rwtxn.commit()?;
     }
+
+    tracing::info!(
+        main_tip = ?main_tip,
+        "Synced {} headers in {:?}",
+        current_height - current_enforcer_tip_height, start.elapsed()
+    );
     Ok(())
 }
 
 // MUST be called after `sync_headers`.
-async fn sync_blocks<MainClient>(
+#[tracing::instrument(skip_all)]
+async fn sync_blocks<MainRpcClient>(
     dbs: &Dbs,
     event_tx: &Sender<Event>,
-    main_client: &MainClient,
+    main_rpc_client: &MainRpcClient,
     main_tip: BlockHash,
 ) -> Result<(), error::Sync>
 where
-    MainClient: bip300301::client::MainClient + Sync,
+    MainRpcClient: bip300301::client::MainClient + Sync,
 {
+    let start = Instant::now();
     let missing_blocks: Vec<BlockHash> = tokio::task::block_in_place(|| {
         let rotxn = dbs.read_txn()?;
-        dbs.block_hashes
+        let ancestor_headers = dbs
+            .block_hashes
             .ancestor_headers(&rotxn, main_tip)
-            .map(|(block_hash, _header)| Ok(block_hash))
-            .take_while(|block_hash| Ok(!dbs.block_hashes.contains_block(&rotxn, block_hash)?))
-            .collect()
-            .map_err(error::Sync::from)
+            .collect::<Vec<_>>()?;
+
+        tracing::info!(
```

Comment:
Can't you just get the block height? IIUC the headers DB always contains all ancestors of any block hash that it contains

Reply:
e3440073d1d8176e2e74b62ceab5de5cf71a52f7